### PR TITLE
Feat list insert and perf improvements rolled into a proper pr

### LINKED ISSUE #360: add `list.insert` function to add elements at an arbitrary index of `list`
`list->insert(index, value)` should add a new element `value` to `list` at index `index`, or cause an error if the `index` is out of bounds.

### PR DESCRIPTION
**Pull request by DarthUdp**
potentially closes #360

### CODE REVIEW

**Comment by DarthUdp** [comment]
after we settle on the implementation it will need tests but I'm not sure I will have the time to implement them I hacked this implementation together in abt an hour in my breaks from work :P

**Comment by DarthUdp** [comment]
on a naive hand checking `list->insert` works as I would expect it to and has the following form
```
list->insert(idx, val)
```

**Comment by CoffeeTableEspresso** [comment]
Yup no worries about the tests, I can do those.

File: interpreter/list_methods.c [hunk 1]
```diff
@@ -430,3 +430,29 @@ int list_sort(struct YASL_State *S) {
...
+	struct YASL_List *new_ls = (struct YASL_List *) new_list->data;
+	// Mutate the list mostly in place
+	ls->count += 1;
+	FOR_LIST(i, elmt, ls) {
+		if (i == (size_t) index) {
+			YASL_List_append(new_ls, value);
+		}
+		YASL_List_append(new_ls, elmt);
+	}
+	ls->items = new_ls->items;
```
Comment (DarthUdp) [inline interpreter/list_methods.c:455]: I suspect a dangling pointer can be left from the old list, or am I overthinking it

**Review by DarthUdp — COMMENTED** [review]
line 455 specifically

File: interpreter/list_methods.c (hunk 1 above)
Comment (CoffeeTableEspresso) [reply interpreter/list_methods.c:455]: Definitely not overthinking. Tests will catch it though, let me push some to your branch.

**Comment by CoffeeTableEspresso** [comment]
This all looks good to me, I'm gonna merge this into the repo, add tests, then merge to v0.11.3.

Thanks for the help on this!

**Comment by DarthUdp** [comment]
One thing I noticed with my last pr my name in the commits went missing and I'm not listed as having contributed

**Comment by DarthUdp** [comment]
Github pr sys can be a bit janky sometimes

**Comment by CoffeeTableEspresso** [comment]
Huh, I've been squash-and-merging commits to master for the past while to keep a cleaner history. 

https://github.blog/changelog/2019-12-19-improved-attribution-when-squashing-commits/ says that should also keep your contributions but apparently not...

I'll try just merging this one and seeing if that does it.

**Comment by DarthUdp** [comment]
> Huh, I've been squash-and-merging commits to master for the past while to keep a cleaner history.
> 
> https://github.blog/changelog/2019-12-19-improved-attribution-when-squashing-commits/ says that should also keep your contributions but apparently not...
> 
> I'll try just merging this one and seeing if that does it.

it usually does, see SnowflakeOS a repo I made contribs for despite squashing it kept my name so idk what happened to github lol

**Comment by CoffeeTableEspresso** [comment]
Ok, then in that case I'll just squash and merge again?

**Pull request merged**